README.md

```diff
@@ -2,54 +2,123 @@
 
   ports = [
     {
-      "address" = "00:bb:4a:d0:5e:38"
-      "pxe_enabled" = "true"
-    }
+      pxe_enabled = true
+      address     = "00:bb:4a:d0:5e:38"
+    },
   ]
-
+  
   properties {
-    "local_gb" = "50"
-    "cpu_arch" =  "x86_64"
+    cpu_arch = "x86_64"
+    local_gb = 50
   }
 
-  instance_info = {
-    "image_source" = "http://172.22.0.1/images/redhat-coreos-maipo-latest.qcow2"
-    "image_checksum" = "26c53f3beca4e0b02e09d335257826fd"
-    "root_gb" = "25"
-    "root_device" = "/dev/vda"
+  root_device {
+   name = "/dev/vda"
   }
-
+   
   driver = "ipmi"
+  
   driver_info {
-			"ipmi_port"=      "6230"
-			"ipmi_username"=  "admin"
-			"ipmi_password"=  "password"
-			"ipmi_address"=   "192.168.111.1"
-			"deploy_kernel"=  "http://172.22.0.1/images/ironic-python-agent.kernel"
-			"deploy_ramdisk"= "http://172.22.0.1/images/ironic-python-agent.initramfs"
+    ipmi_port       =  "6230"
+    ipmi_username   =  "admin"
+    ipmi_password   =  "password"
+    ipmi_address    =  "192.168.111.1"
+    deploy_kernel   =  "http://172.22.0.1/images/ironic-python-agent.kernel"
+    deploy_ramdisk  =  "http://172.22.0.1/images/ironic-python-agent.initramfs"
   }
 }
 ```
 
+## Ports
+
+Ports may be specified as part of the node resource, or as a separate `ironic_port_v1`
+declaration.
+
+
+```
+resource "ironic_port_v1" "openshift-master-0-port-0" {
+  node_uuid   = "${ironic_node_v1.openshift-master-0.id}"
+  pxe_enabled = true
+  address     = "00:bb:4a:d0:5e:38"
+}
+```
+
+## Allocation
+
+```
+resource "ironic_allocation_v1" "openshift-master-allocation" {
+  count = 3
+
+  resource_class = "baremetal"
+    
+  candidates = [
+    "${ironic_node_v1.openshift-master-0.id}",
+    "${ironic_node_v1.openshift-master-1.id}",
+    "${ironic_node_v1.openshift-master-2.id}",
+  ]
+  
+  traits = [
+    "CUSTOM_FOO",
+  ]
+}
+```
+
+## Deployment
+
+A deployment will provision a baremetal node, using the information given in the resource.  The `count` metaparameter
+can be used to deploy multiple nodes. Terraform will drive the [Ironic state machine](https://docs.openstack.org/ironic/latest/contributor/states.html) to bring the node to an `active` state.
+It operates similar to a [Metalsmith deployment](https://docs.openstack.org/metalsmith/latest/), where a deployment
+consumes a node.  This allows a user to manage their hardware outside of the provider if desired.
+
+If the `allocation` parameter is set, an Ironic allocation will be used to select a node.  Otherwise, a user may specify
+a specific `node_uuid` to use.
+
+Destruction of a deployment cleans a node and brings it back to the `available` state.
+
+```
+resource "ironic_deployment" "masters"  {
+  count = 3
+  allocation = "${ironic_allocation_v1.openshift-master-allocation[$count.index]}"
+ 
+  user_data      = "${file("master.ign")}"
+  image_source   = "http://172.22.0.1/images/redhat-coreos-maipo-latest.qcow2"
+  image_checksum = "26c53f3beca4e0b02e09d335257826fd"
```

Comment:
For partition images (which metalkube does not use) you also need `image_kernel`, `image_ramdisk` and `root_gb`. All of these go to `instance_info` as well.